flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java

```diff
@@ -0,0 +1,810 @@
+                                            // elementWindows
+                                            return false; // isSkippedElement = false
+                                        }));
+            }
+
+            // side output input event if
+            // element not handled by any window
+            // late arriving tag has been set
+            // windowAssigner is event time and current timestamp + allowed lateness no less than
+            // element timestamp
+            StateFutureUtils.combineAll(windowFutures)
+                    .thenAccept(
+                            isSkippedElements -> {
+                                boolean isSkippedElement = true;
+                                for (Boolean isSkipped : isSkippedElements) {
+                                    if (!isSkipped) {
+                                        isSkippedElement = false;
+                                        break;
+                                    }
+                                }
+
+                                if (isSkippedElement && isElementLate(element)) {
+                                    if (lateDataOutputTag != null) {
+                                        sideOutput(element);
+                                    } else {
+                                        this.numLateRecordsDropped.inc();
+                                    }
+                                }
+                            });
+        }
+    }
+
+    @Override
+    public void onEventTime(InternalTimer<K, W> timer) throws Exception {
+        windowDeclaredVariable.set(timer.getNamespace());
+        AtomicReference<TriggerResult> triggerResult = new AtomicReference<>();
+
+        if (windowAssigner instanceof MergingWindowAssigner) {
+            throw new UnsupportedOperationException(
+                    "Async WindowOperator not support merging window (e.g. session window) yet.");
+        } else {
+            triggerContext
+                    .onEventTime(timer.getTimestamp())
+                    .thenAccept(triggerResult::set)
+                    .thenConditionallyCompose(
+                            ignore -> triggerResult.get().isFire(),
+                            ignore ->
+                                    windowState
+                                            .asyncGet()
+                                            .thenConditionallyCompose(
+                                                    Objects::nonNull,
+                                                    contents ->
+                                                            emitWindowContents(
+                                                                    timer.getKey(),
+                                                                    timer.getNamespace(),
+                                                                    contents),
+                                                    StateFutureUtils.createVoidReturnFunction()),
```

Comment:
Consider import static `StateFutureUtils.createVoidReturnFunction` since you are using this so many times.

Reply:
Or you can use the `thenConditionallyCompose` with only true branch provided.

Reply:
Using `thenConditionallyCompose ` is a great idea, I have transformed to `thenConditionallyCompose` with only true branch.

 `StateFutureUtils.createVoidReturnFunction` has been removed.